pkg/server/image_pull.go

```diff
@@ -17,14 +17,202 @@ limitations under the License.
+	var meta *metadata.ImageMetadata
+	meta, err = c.imageMetadataStore.Get(digest)
+	if err == nil {
+		if meta == nil {
+			meta = &metadata.ImageMetadata{
+				ID:          digest,
+				RepoTags:    []string{image},       // TODO fill this array properly
+				RepoDigests: []string{digest},      // TODO fill this array properly
+				Size:        uint64(size),          // TODO  compressed or uncompressed size? using compressed
+				Auth:        r.GetAuth(),           // TODO security
+				CreatedAt:   time.Now().UnixNano(), // TODO or should this be the image created at time?
+			}
+			if err = c.imageMetadataStore.Create(*meta); err != nil {
+				return &runtime.PullImageResponse{ImageRef: digest},
+					fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+			}
+		}
+	}
+
+	// Return the image digest
+	return &runtime.PullImageResponse{ImageRef: digest}, err
+}
+
+// imageReferenceResolver attempts to resolve the image reference into a name
+// and manifest via the containerd library call..
+//
+// The argument `ref` should be a scheme-less URI representing the remote.
+// Structurally, it has a host and path. The "host" can be used to directly
+// reference a specific host or be matched against a specific handler.
+//
+// The returned name should be used to identify the referenced entity.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+//
+// If the resolution fails, an error will be returned.
+func (c *criContainerdService) imageReferenceResolver(ref string) (resolvedImageName string, manifest imagespec.Manifest, compressedSize uint64, err error) {
+	var (
+		ctx  = gocontext.Background()
+		size int64
+		desc imagespec.Descriptor
+	)
+
+	// This first go routine in the local error group eg will resolve the image
+	// name; place that in the image store with another go routine; then dispatch
+	// a handler to fetch the object the manifest
+	resolver := docker.NewResolver()
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		var (
+			err     error
+			fetcher remotes.Fetcher
+		)
+
+		resolvedImageName, desc, fetcher, err = resolver.Resolve(ctx, ref)
+		if err != nil {
+			return err
+		}
+
+		eg.Go(func() error {
+			return c.imageStore.Put(ctx, resolvedImageName, desc)
+		})
+
+		return containerd_images.Dispatch(
+			ctx,
+			remotes.FetchHandler(c.ingester, fetcher),
+			desc)
+	})
+
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		errs <- eg.Wait()
+	}()
+
+	defer func() {
+		nctx := gocontext.Background()
+		image, err := c.imageStore.Get(nctx, resolvedImageName)
+		if err != nil {
+			fmt.Printf("\nc.imageStore.Get for %q ERR: %v\n", resolvedImageName, err)
+		}
+		p, err := content.ReadBlob(nctx, c.provider, image.Target.Digest)
+		if err != nil {
+			fmt.Printf("\ncontent.ReadBlob ERR: %v\n", err)
+		}
+		err = json.Unmarshal(p, &manifest)
```

Comment:
We need image config not manifest for now. https://github.com/containerd/containerd/blob/master/images/image.go#L29